pennylane/devices/qubit/adjoint_jacobian.py

```diff
@@ -0,0 +1,167 @@
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Functions to apply adjoint jacobian differentiation"""
+
+import warnings
+import numpy as np
+
+import pennylane as qml
+
+from pennylane.operation import operation_derivative
+from pennylane.tape import QuantumTape
+from pennylane.measurements import ExpectationMP
+
+from .apply_operation import apply_operation
+
+from .initialize_state import create_initial_state
+
+# pylint: disable=protected-access, too-many-branches
+
+
+def _dot_product_real(bra, ket, num_wires):
+    """Helper for calculating the inner product for adjoint differentiation."""
+    # broadcasted inner product not summing over first dimension of the bra tensor
+    sum_axes = tuple(range(1, num_wires + 1))
+    return qml.math.real(qml.math.sum(qml.math.conj(bra) * ket, axis=sum_axes))
+
+
+def adjoint_jacobian(
+    tape: QuantumTape, prep_operation: qml.operation.StatePrep = None
+):  # pylint: disable=too-many-statements
+    """Implements the adjoint method outlined in
+    `Jones and Gacon <https://arxiv.org/abs/2009.02823>`__ to differentiate an input tape.
+
+    After a forward pass, the circuit is reversed by iteratively applying adjoint
+    gates to scan backwards through the circuit.
+
+    .. note::
+        The adjoint differentiation method has the following restrictions:
+
+        * As it requires knowledge of the statevector, only statevector simulator devices can be
+          used.
+
+        * Only expectation values are supported as measurements.
+
+        * Does not work for parametrized observables like :class:`~.Hamiltonian`,
+          :class:`~.Sum` or :class:`~.Hermitian`.
+
+    Args:
+        tape (.QuantumTape): circuit that the function takes the gradient of
+        prep_operation (.StatePrep): state preparation operation applied for state initialization
+
+    Returns:
+        array or tuple[array]: the derivative of the tape with respect to trainable parameters.
+        Dimensions are ``(len(observables), len(trainable_params))``.
+
+    Raises:
+        QuantumFunctionError: if the input tape has measurements that are not expectation values
+            or contains a multi-parameter operation aside from :class:`~.Rot`
+    """
+
+    # Check validity of measurements
+    for m in tape.measurements:
+        if not isinstance(m, ExpectationMP):
+            raise qml.QuantumFunctionError(
+                "Adjoint differentiation method does not support"
+                f" measurement {m.__class__.__name__}"
+            )
+
+        if m.obs.name in {"Hamiltonian"}:
+            raise qml.QuantumFunctionError(
+                f"Adjoint differentiation method does not support observable {m.obs.name}."
+            )
```

Comment:
We can move these checks to `preprocess`.  If someone requests adjoint differentiation through the execution config, we can run these checks. So they don't need to exist here.

Reply:
Maybe we could pull all the preprocessing into a separate function located in this file, `adjoint_jacobian_preprocessing`.  Then the preprocessing code would be located close to execution code, but called at a different stage in the pipeline.

Reply:
I think that could lead to some unnecessary errors. If someone specifies `diff_method="adjoint"` when creating a QNode but don't actually run `qml.grad` or `qml.jacobian`, an error shouldn't be raised.